project-koku/koku
# [COST-4620] - Add basic provider data validation

## Jira Ticket

[COST-4620](https://issues.redhat.com/browse/COST-4620)

## Description

- Add a new post summary task to validate some total metrics between postgres and trino data. 
- Logs if data looks valid or not and marks the provider with a true/false accordingly (Giving us a quick way to build a list of incorrect provider data)
- Additionally adds a trigger for OCP on Cloud clusters too.
- New Masu endpoint to manually trigger this validation for a given date range.
- Also includes an unleash flag so we can skip running this task by default and gradually enable customers. Just to be aware of any performance hits.


## Testing

1. Checkout Branch
2. Restart Koku
3. Enable the validation unleash flag `is_validation_enabled`
4. Ingest data like normal.
5. You should see logging as follows per provider ingested
```
{'message': 'validation started for provider: a67bb4a2-8e05-40
{'message': 'triggering VALIDATION_QUERY',
{'message': 'finished VALIDATION_QUERY',
{'message': 'executing trino sql', 'tracing_id': '', 'log_ref': 'data validation query'}
{'message': 'executed trino sql', 'tracing_id': '', 'log_ref': 'data validation query', 'running_time': 0.5209639072418213}
{'message': 'all data complete for provider: a67bb4a2-8e05-4078-b1e4-9c8525f3fb0f'
```
6. You can also re-trigger the validation task via this masu endpoint: 
            http://localhost:5042/api/cost-management/v1/validate_cost_data/?provider_uuid={uuid}&start_date=2024-07-01&end_date=2024-07-16
7. You should see similar to the above in the logs

## Testing invalid data
1. After you have ingest some data lets mess it up in postgres!
2. Heres a command for AWS:
            `update org1234567.reporting_awscostentrylineitem_daily_summary set unblended_cost = 0 where product_code = 'AmazonEC2';`
3. Run validation via masu again
4. see the following log messages
```
'message': 'validation started for provider: a67bb4a2-8e05-4078-b1e4-9c8525f3fb0f'
'message': 'triggering VALIDATION_QUERY'
'message': 'finished VALIDATION_QUERY'
'message': 'executing trino sql'
'message': 'executed trino sql'
'message': "provider has incomplete data for specified days: {datetime.date(2024, 7, 1): {'pg_value': 3789.255140293, 'trino_value': 6347.878609799311, 'delta': 2558.623469506311},
 ```

## Testing OCP on Cloud clusters
1. Ingest OCP data
2. Ingest correlated cloud data
3. Check logs for matching validation for the correlated clusters similar to those above.

## Testing via masu
1. Ingest data
2. Hit the following endpoint `http://localhost:5042/api/cost-management/v1/validate_cost_data/?provider_uuid={provider_uuid}&start_date=2024-07-01&end_date=2024-07-19` (Sub your provider ID)
3. See validation log messages.

## Testing an OCP on CLOUD cluster via masu
1. Ingest ocp on cloud data
2. Hit the following endpoint `http://localhost:5042/api/cost-management/v1/validate_cost_data/?provider_uuid={provider_uuid}&ocp_on_cloud_type={type}&start_date=2024-07-01&end_date=2024-07-19` (Sub your provider ID)
3. See validation log messages similar to above

## Notes:

Along side the testing about if you want to verify the query data you can enable `DEBUG` logs and you should see the comparison data from Trino/PG logged for quick comparison. 
Example: [2024-07-22 12:45:54,311] INFO fc9a26e0-a60a-4264-8aac-afebb4f082e5 3268 PG: {datetime.date(2024, 7, 18): 3870.034663868, datetime.date(2024, 7, 19): 3670.98312068} Trino data: {datetime.date(2024, 7, 18): 3870.034663795944, datetime.date(2024, 7, 19): 3670.9831205979876}

## Release Notes

```markdown
* [COST-4620](https://issues.redhat.com/browse/COST-4620) Add internal data validation
```
